src/hotspot/share/gc/shared/memAllocator.cpp

```diff
@@ -408,11 +409,28 @@ oop ObjArrayAllocator::initialize(HeapWord* mem) const {
   assert(_length >= 0, "length should be non-negative");
   if (_do_zero) {
     mem_clear(mem);
+    mem_zap_end_padding(mem);
   }
   arrayOopDesc::set_length(mem, _length);
   return finish(mem);
 }
 
+#ifndef PRODUCT
+void ObjArrayAllocator::mem_zap_end_padding(HeapWord* mem) const {
+  const size_t length_in_bytes = static_cast<size_t>(_length) << ArrayKlass::cast(_klass)->log2_element_size();
+  const BasicType element_type = ArrayKlass::cast(_klass)->element_type();
+  const size_t base_offset_in_bytes = arrayOopDesc::base_offset_in_bytes(element_type);
+  const size_t size_in_bytes = _word_size * BytesPerWord;
+
+  const address obj_end = reinterpret_cast<address>(mem) + size_in_bytes;
+
+  const size_t padding_in_bytes = size_in_bytes - base_offset_in_bytes - length_in_bytes;
```

Comment:
Since this is unsigned subtraction, could you add a assert that `size_in_bytes` is larger or equal? (Or calculate `zap_start = obj_start + offset + length` and compare it with `obj_end`.)